src/coreclr/jit/codegencommon.cpp

```diff
@@ -10801,6 +10806,102 @@ void CodeGen::genIPmappingGen()
     compiler->eeSetLIdone();
 }
 
+#ifdef DEBUG
+void CodeGen::genDumpPreciseDebugInfoInlineTree(FILE* file, InlineContext* context, bool* first)
+{
+    if (context->GetSibling() != nullptr)
+    {
+        genDumpPreciseDebugInfoInlineTree(file, context->GetSibling(), first);
+    }
+
+    if (context->IsSuccess())
+    {
+        if (!*first)
+        {
+            fprintf(file, ",");
+        }
+
+        *first = false;
+
+        fprintf(file, "{\"Ordinal\":%u,", context->GetOrdinal());
+        fprintf(file, "\"MethodID\":%lld,", (INT64)context->GetCallee());
+        const char* className;
+        const char* methodName = compiler->eeGetMethodName(context->GetCallee(), &className);
+        fprintf(file, "\"MethodName\":\"%s\",", methodName);
+        fprintf(file, "\"Inlinees\":[");
+        if (context->GetChild() != nullptr)
+        {
+            bool childFirst = true;
+            genDumpPreciseDebugInfoInlineTree(file, context->GetChild(), &childFirst);
+        }
+        fprintf(file, "]}");
+    }
+}
+
+void CodeGen::genDumpPreciseDebugInfo()
+{
+    if (JitConfig.JitDumpPreciseDebugInfoFile() == nullptr)
+        return;
+
+    static unsigned int s_flag;
+
+    while (InterlockedCompareExchange(&s_flag, 1, 0) != 0)
+        System_YieldProcessor();
```

Comment:
Do we have a better lock primitive in the JIT I can use here?

Reply:
There is [`CritSecHolder`](https://github.com/dotnet/runtime/blob/7afb66262c287b7832b872473d1e668038aa2496/src/coreclr/jit/utils.h#L741), if that would be suitable here.

Reply:
That looks better, thanks.